Source/WTF/wtf/text/StringSearch.h

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <limits>
+#include <wtf/Span.h>
+#include <wtf/text/StringCommon.h>
+#include <wtf/text/StringView.h>
+
+namespace WTF {
+
+template<typename OffsetType>
+class BoyerMooreHorspoolTable {
+    WTF_MAKE_FAST_ALLOCATED(BoyerMooreHorspoolTable);
+public:
+    static constexpr unsigned size = 256;
+    static constexpr unsigned maxPatternLength = std::numeric_limits<OffsetType>::max();
+
+    explicit BoyerMooreHorspoolTable(StringView pattern)
+    {
+        size_t length = pattern.length();
+        ASSERT(length <= maxPatternLength);
+        if (length) {
+            for (auto& element : m_table)
+                element = length;
+            for (unsigned i = 0; i < (pattern.length() - 1); ++i) {
+                unsigned index = pattern[i] & 0xff;
+                m_table[index] = length - 1 - i;
+            }
+        }
+    }
+
+    explicit constexpr BoyerMooreHorspoolTable(ASCIILiteral pattern)
+    {
+        size_t length = pattern.length();
+        ASSERT(length <= maxPatternLength);
+        if (length) {
+            for (auto& element : m_table)
+                element = length;
+            for (unsigned i = 0; i < (pattern.length() - 1); ++i) {
+                unsigned index = pattern.characters()[i] & 0xff;
+                m_table[index] = length - 1 - i;
+            }
+        }
+    }
```

Comment:
These 2 constructors are largely identical.  Can we add a template utility function and have the 2 constructors call that instead?

Reply:
Changed.